plugins/module_utils/system_access_users_utils.py

```diff
@@ -359,36 +420,43 @@ def _generate_hashed_secret(self, secret: str) -> str:
             """
         )
 
-    def set_apikeys(self, apikeys: list = None) -> list:
+    @staticmethod
+    def generate_apikeys(apikeys: List[dict] = None) -> List[dict]:
         """
-        Generates a list of dictionaries, each containing a 'key' and a 'secret'.
-        If apikeys is provided, each element in apikeys is used as the 'key',
-        and a new 'secret' is generated.
-        If apikeys is not provided or is an empty list, a single 'key'-'secret' pair is generated.
+        Generates API keys if they are missing or validates provided keys.
 
         Args:
-            apikeys (list, optional): A list of strings to be used as the 'key' part of each pair.
+            apikeys (list[dict]): A list of dictionaries containing 'key' and 'secret'.
 
         Returns:
-            list: A list of dictionaries, where each dictionary has a 'key' and a 'secret'.
+            list[dict]: A list of dictionaries with valid 'key' and 'secret' pairs.
+
+        Raises:
+            OPNSenseNotValidBase64APIKeyError: If the provided key or secret
+            is not a valid base64 string.
         """
 
-        api_keys = []
+        api_keys: list[dict] = []
 
-        # Check if apikeys is None or contains an empty string
-        if apikeys is None or "" in apikeys:
-            key = base64.b64encode(os.urandom(60)).decode("utf-8")
-            secret = base64.b64encode(os.urandom(60)).decode("utf-8")
-            api_keys.append({"key": key, "secret": secret})
-        else:
-            for api_key in apikeys:
-                secret = base64.b64encode(os.urandom(60)).decode("utf-8")
+        for apikey in apikeys:
+            # Check if key and secret are provided
+            if not apikey["key"]:
+                key = base64.b64encode(os.urandom(60)).decode("utf-8")
+
+                if not apikey["secret"]:
+                    secret = base64.b64encode(os.urandom(60)).decode("utf-8")
+
+                api_keys.append({"key": key, "secret": secret})
+            else:
                 try:
-                    base64.b64decode(api_key)
-                    api_keys.append({"key": api_key, "secret": secret})
+                    base64.b64decode(apikey["key"])
+                    base64.b64decode(apikey["secret"])
+
+                    api_keys.append(apikey)
+
```

Comment:
Is there any validation at any point for this case:
```
key : SOME_VALID_KEY
secret: 
```

With a key provided but no secret provided, this function proceds to line 453 to base64 encode an empty secret. 

I have tried this test case:
```yaml
    # Test User with valid api_keys
    - name: "Test User 21: Test User Creation with valid api_keys"
      puzzle.opnsense.system_access_users:
            username: test_user_21
            password: test_password_21
            apikeys:
              - key: "TEST_API_KEY_WITH_RANDOM_CHARS_UNTIL_80_zo5Y3bUpOQFfbQnAOB6GqbHsPAP9Jqbjofnqu900"
                secret: ""
            full_name: "Test User 21: Test User Creation with valid api_keys"
      register: api_keys_result

    - name: Return the created apikeys and secret of Test User 21
      ansible.builtin.debug:
        msg: "The following api_keys were created {{ api_keys_result.generated_apikeys }}"
      when:
        - "'generated_apikeys' in api_keys_result"
        - api_keys_result.generated_apikeys | length > 0
``` 

Which results in this output:
```

TASK [Test User 21: Test User Creation with valid api_keys] ********************
changed: [23.7]

TASK [Return the created apikeys and secret of Test User 21] *******************
ok: [23.7] => {
    "msg": "The following api_keys were created [{'key': 'TEST_API_KEY_WITH_RANDOM_CHARS_UNTIL_80_zo5Y3bUpOQFfbQnAOB6GqbHsPAP9Jqbjofnqu900', 'secret': ''}]"
}

PLAY RECAP *********************************************************************
23.7                       : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

```

Is this behaviour desired or should this method check also for an empty secret?

Reply:
This function was revised and will now generate a secret of None is provided.

Reply:
alright 👍🏼